### PR TITLE
ENH: include (fault) surfaces from FaultRoom plugin

### PR DESCRIPTION
**Pull request by jcrivenaes**
Address #362

### CODE REVIEW

File: tests/test_units/test_utils.py
Comment (janbjorge) [inline tests/test_units/test_utils.py:228]: consider using `pytest.approx(...) == ...`

File: src/fmu/dataio/_objectdata_provider.py
Comment (janbjorge) [inline src/fmu/dataio/_objectdata_provider.py:485]: Consider
```suggestion
        return (
             isinstance(self.obj, dict)
            and "metadata" in self.obj
            and "source" in self.obj["metadata"]
            and "FaultRoom" in self.obj["metadata"]["source"]
```

File: examples/s/d/nn/xcase/realization-0/iter-0/rms/bin/export_faultroom_surfaces.py
Comment (janbjorge) [inline examples/s/d/nn/xcase/realization-0/iter-0/rms/bin/export_faultroom_surfaces.py:37]: This is so nice, love it!

File: src/fmu/dataio/readers.py
Comment (janbjorge) [inline src/fmu/dataio/readers.py:22]: Potential talking point: the builtin json parser is a bit "slow" compared to parsers like https://docs.rs/jiter/latest/jiter/ . Given the size of these files, it might be worth a bit of looking into (at a later stage)

File: src/fmu/dataio/readers.py
Comment (janbjorge) [inline src/fmu/dataio/readers.py:86]: nit, use min(...), max(...)

File: src/fmu/dataio/readers.py
Comment (janbjorge) [inline src/fmu/dataio/readers.py:76]:
```suggestion
        xmin = ymin = zmin = float('inf')
        xmax = ymax = zmax = float('-inf')
```

File: examples/s/d/nn/xcase/realization-0/iter-0/rms/bin/export_faultroom_surfaces.py
Comment (perolavsvendsen) [inline examples/s/d/nn/xcase/realization-0/iter-0/rms/bin/export_faultroom_surfaces.py:17]: Perhaps confusing that this is pointing to `faultlines` which is not the FaultRoom-output. Suggest something like "MyFaultroomOutput" or similarly generic.

(What does "GL" mean?)

File: examples/s/d/nn/xcase/realization-0/iter-0/rms/bin/export_faultroom_surfaces.py
Comment (perolavsvendsen) [reply examples/s/d/nn/xcase/realization-0/iter-0/rms/bin/export_faultroom_surfaces.py:37]: yes, agree - and this is also very consistent with the established pattern of 1) representing whatever data as a Python object, then 2) feeding that Python object to `.export()`.

File: schema/definitions/0.8.0/examples/surface_faultroom_data.yml
Comment (perolavsvendsen) [inline schema/definitions/0.8.0/examples/surface_faultroom_data.yml:1]: fault surface, faultroom surface, something different than "depth surface"?

File: schema/definitions/0.8.0/examples/surface_faultroom_data.yml
Comment (perolavsvendsen) [inline schema/definitions/0.8.0/examples/surface_faultroom_data.yml:71]: Maybe make filename more similar to actual output (with the short hash, etc)

File: schema/definitions/0.8.0/examples/surface_faultroom_data.yml
Comment (perolavsvendsen) [inline schema/definitions/0.8.0/examples/surface_faultroom_data.yml:92]: ref comment on filename

**Review by perolavsvendsen — COMMENTED** [review]
See comments. Not entirely sure if `class: surface` is the best option, although I don't think it's "wrong". If we merge this, need to clarify with later users of these data. E.g. there may be some assumptions somewhere that `class: surface` == "an IRAP Binary formatted regular surface".

- [ ] Explicitly check with REP
- [ ] Explicitly check with Sumo
- [ ] Generally check with other potential consumers

File: examples/s/d/nn/xcase/realization-0/iter-0/rms/bin/export_faultroom_surfaces.py
Comment (jcrivenaes) [reply examples/s/d/nn/xcase/realization-0/iter-0/rms/bin/export_faultroom_surfaces.py:17]: Copy/paste of other example... will fix

File: schema/definitions/0.8.0/examples/surface_faultroom_data.yml
Comment (jcrivenaes) [reply schema/definitions/0.8.0/examples/surface_faultroom_data.yml:1]: Yes, will fix

File: schema/definitions/0.8.0/examples/surface_faultroom_data.yml
Comment (jcrivenaes) [reply schema/definitions/0.8.0/examples/surface_faultroom_data.yml:71]: good catch

**Comment by jcrivenaes** [comment]
> `class: surface` == "an IRAP

I think it was never the intention that class `surface` is a regular grid (nor a specific file format), so `surface` is more generic. Then the  `format` and `layout` fields are needed by consumer in order to parse in a correct manner

File: src/fmu/dataio/readers.py
Comment (jcrivenaes) [reply src/fmu/dataio/readers.py:22]: We can take it later

**Review by janbjorge — APPROVED** [review]
Looks good to me, i cant comment on the domain related questions.

**Comment by perolavsvendsen** [comment]
Still waiting for feedback from REP, Webviz, DynaGeo and others who may work on the assumption that `class: surface` == `An IRAP Binary formatted file`.

**Comment by jcrivenaes** [comment]
> Still waiting for feedback from REP, Webviz, DynaGeo and others who may work on the assumption that `class: surface` == `An IRAP Binary formatted file`.

Any progress here?

**Comment by perolavsvendsen** [comment]
> Any progress here?

Some... But I don't think we have to wait for _everything_ to be ready _everywhere_. I think we just merge this PR if we are happy.

**Review by perolavsvendsen — APPROVED** [review]
I say we merge this.

File: src/fmu/dataio/providers/objectdata/_faultroom.py
Comment (mferrera) [inline src/fmu/dataio/providers/objectdata/_faultroom.py:26]: I think the `assert` should not be necessary with `obj: FaultRoomSurface` annotated in the class

File: src/fmu/dataio/providers/objectdata/_faultroom.py
Comment (mferrera) [inline src/fmu/dataio/providers/objectdata/_faultroom.py:43]: Same here with the assert -- does mypy still complain with the assert removed, but the type hint in the class?

File: src/fmu/dataio/readers.py
Comment (mferrera) [inline src/fmu/dataio/readers.py:21]: nit: specifying `"r"` isn't necessary, it is the default

File: src/fmu/dataio/readers.py
Comment (mferrera) [inline src/fmu/dataio/readers.py:73]: `_get` to me implies the method returns something; maybe `_set` might be more accurate`

**Review by mferrera — APPROVED** [review]
Some comments for your consideration but LGTM. I cannot comment on the fault room specifics, of course

File: src/fmu/dataio/dataio.py
Comment (tnatt) [inline src/fmu/dataio/dataio.py:716]: I am a bit unsure of how the object can change class along the way, and the need for this self._object.
Can you eloborate?

Also I think we should change the `self._check_obj_if_file(obj)` method so that it doesn't return anything. Right now it is a bit misleading as it appears that the `obj` is being modified in there, but it is not. We can change it in another PR 🙂

Comment (tnatt) [reply src/fmu/dataio/dataio.py:716]: In the `generate_export_metadata` function below we take in the `obj` and not the `self._object`, it should probably be changed there as well to be consistent 👍

File: src/fmu/dataio/readers.py
Comment (jcrivenaes) [reply src/fmu/dataio/readers.py:73]: Good catch; changed it

**Pull request merged**
